### DISCUSSION TITLE
Simulation interrupted: -1: Could not insert V2X message into network

### DISCUSSION THREAD

**Discussion by SafrasIqbal** [eclipse/mosaic#306] [Q&A] [answered]
Hello Everyone.

Need help fixing an issue in my simulation. The simulation is stopped halfway with this error message

`2023-04-03 12:00:18,938 INFO  SequentialTimeManagement - Simulation interrupted: -1
2023-04-03 12:00:21,510 ERROR MosaicStarter - Stopping simulation due to a critical error.
org.eclipse.mosaic.rti.api.InternalFederateException: Could not insert V2X message into network.
	at org.eclipse.mosaic.lib.coupling.AbstractNetworkAmbassador.process(AbstractNetworkAmbassador.java:605)
	at org.eclipse.mosaic.lib.coupling.AbstractNetworkAmbassador.processInteraction(AbstractNetworkAmbassador.java:290)
	at org.eclipse.mosaic.rti.api.AbstractFederateAmbassador.advanceTime(AbstractFederateAmbassador.java:108)
	at org.eclipse.mosaic.rti.time.SequentialTimeManagement.runSimulation(SequentialTimeManagement.java:90)
	at org.eclipse.mosaic.starter.MosaicSimulation.runSimulation(MosaicSimulation.java:188)
	at org.eclipse.mosaic.starter.MosaicStarter.execute(MosaicStarter.java:132)
	at org.eclipse.mosaic.starter.MosaicStarter.main(MosaicStarter.java:74)
Caused by: org.eclipse.mosaic.rti.api.IllegalValueException: Node not simulated: veh_1
	at org.eclipse.mosaic.lib.coupling.AbstractNetworkAmbassador.process(AbstractNetworkAmbassador.java:601)
	... 6 common frames omitted`

It seems vehicle get removed when its reached the destination before the simulation ends. Therefore, V2X message expected to delivered to the vehicle not in the network, this cause the critical error. I don't know how to resolve this. Thank you in advance for the help on this.   

Regards,
Safras

**Comment by schwepmo** [comment]
Hello,
from your question alone it is hard to figure out the reason for the issue.
Could you tell us what network simulator you are using and how you are trying to send the message?

Vehicles are indeed removed from simulation once they finish their routes. You won't be able to address a vehicle once it finishes its route.
If you need the vehicle to stay in the simulation for the entire duration of you simulation you could try to manually keep it simulated by stopping the vehicle at the end of its route.

**Reply by SafrasIqbal** [reply]
Hello @schwepmo ,

Thank you for your prompt response. 

I am using OMNET++ for vehicular communication and SUMO as it default. Sending CAM messages using VehicleCamSendingApp class. 

On the second thought, vehicle seems to be removed before its reach the destination due to a critical error , but I cannot find what is the error causing this issue.  I also have other scenarios works with VehicleCamSendingApp class without any issue. 

Please let me know any other information you need to check this issue. Thank you.  

Regards,
Safras

**Reply by schwepmo** [reply]
Hey Safras,
I recommend looking into the log-files generated by MOSAIC, especially *MOSAIC.log* and *Traffic.log* to get insights as to why the vehicle is removed. Your issue typically suggests that something goes wrong in the traffic simulator.
How are you defining your vehicles, via the *mapping_config.json* or using a SUMO route-file?

**Reply by SafrasIqbal** [reply]
Hello @schwepmo , 
I do insert the vehicle using mapping json file and applications attached here too. 

I checked again the scenario visual. Issue actually occurs when the first vehicle reaches the destination. The error thrown from this package - org.eclipse.mosaic.rti.api.InternalFederateException: , 
Do you think its worth to add a condition in here if vehicle exist send the message otherwise do nothing.   Appreciate your suggestion.

also MOSAIC logs look like this, 
  
![MOSAIC_1](https://user-images.githubusercontent.com/18607009/229791891-d1f7fd3e-e913-4bea-b25c-4407797412f3.png)

and the traffic log : 
![Traffic](https://user-images.githubusercontent.com/18607009/229793903-ca39970b-c004-4d24-98da-a90939b87ac9.png)

**Reply by schwepmo** [reply]
From the *Traffic.log* it looks like your traffic simulation runs fine.
The issue really boils down to your addressed vehicle (veh_1) already being removed from simulation because it finishes it's route.

I wouldn't suggest adding the check as this isn't trivial and also not applicable to the real world, because you cannot know which vehicles are reachable within an VANET.

However, this really seems to be a bug on our side as the veh_1 shouldn't be addressed by the CAM after leaving the simulation.
Would you mind attaching your zipped scenario, so we can try to reproduce the error?

**Reply by kschrab** [reply, marked as answer]
I guess this could also be just a problem of how the `VehicleCamSendingApp`/`AbstractCamSendingApp` is implemented. The application periodically tries to send a CAM message. In the `sample` method it also checks if the vehicle is already removed from the simulation (`isTornDown`). However, it then adds an event to call another method in the future (`checkDataAndSendCam`) where the actual sending takes place. This method, however, does _not_ check if the vehicle is still alive. So maybe here lies the problem. 

You could also fix this by yourself by adding a `if (isTornDown()) { return; }` at the beginning of the `checkDataAndSendCam` method.

**Reply by SafrasIqbal** [reply]
Hi @kschrab and @kschrab ,

Thank you for the solution , it works like charm :+1: . Appreciate the prompt support.

Regards,
Safras

**Reply by kschrab** [reply]
Thanks for the finding, we will definitely remove this trap in the next release 😄
